File: core/network/src/main/java/app/muplay/network/CapabilityNegotiator.java
package app.muplay.network;

import app.muplay.model.ServerCapabilities;
import com.google.common.util.concurrent.FluentFuture;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.MoreExecutors;
import java.io.IOException;
import java.util.Map;
import javax.annotation.Nonnull;

/**
 * Three-tier capability negotiation: {@code ping} establishes reachability and whether the
 * server speaks OpenSubsonic at all; only then is it worth asking for the extension list.
 *
 * <p>The extension fetch itself is allowed to fail without failing negotiation: a server that
 * advertised {@code openSubsonic=true} on {@code ping} but errors on {@code
 * getOpenSubsonicExtensions} (an older OpenSubsonic-flagged server that predates the endpoint, a
 * transient 404, or a Subsonic-level error in the body) still degrades to a capabilities object
 * with no extensions rather than failing every caller of {@link #negotiate()}. {@link
 * SubsonicClient} maps both an unsuccessful HTTP response and a Subsonic-level error onto an
 * {@link IOException} (the latter via {@link SubsonicErrorException}), so catching {@link
 * IOException} here covers both without masking an unrelated programming error.
 */
public final class CapabilityNegotiator {

  private final SubsonicClient client;

  public CapabilityNegotiator(@Nonnull SubsonicClient client) {
    this.client = client;
  }

  @Nonnull
  public ListenableFuture<ServerCapabilities> negotiate() {
    return FluentFuture.from(client.ping())
        .transformAsync(
            info -> {
              if (!info.openSubsonic()) {
                return Futures.immediateFuture(ServerCapabilities.NONE);
              }
              return FluentFuture.from(client.getOpenSubsonicExtensions())
                  .transform(
                      exts -> new ServerCapabilities(true, exts), MoreExecutors.directExecutor())
                  .catching(
                      IOException.class,
                      e -> new ServerCapabilities(true, Map.of()),
                      MoreExecutors.directExecutor());
            },
            MoreExecutors.directExecutor());
  }
}

File: core/network/src/main/kotlin/app/muplay/network/CapabilityNegotiator.kt
package app.muplay.network

import app.muplay.model.ServerCapabilities

/**
 * Three-tier OpenSubsonic capability negotiation over a [SubsonicClient]: `ping` first, and only
 * if that reports [app.muplay.model.ServerInfo.isOpenSubsonic] does this attempt
 * `getOpenSubsonicExtensions` at all.
 *
 * The two degraded paths this negotiates around are deliberately different outcomes, not the same
 * one reached two ways:
 *
 * - `ping` reporting `isOpenSubsonic == false` (a plain, legacy Subsonic server) negotiates
 *   straight to [ServerCapabilities] with no extensions, without even attempting
 *   `getOpenSubsonicExtensions` — see [negotiate].
 * - An OpenSubsonic server (`ping` reported `isOpenSubsonic == true`) whose
 *   `getOpenSubsonicExtensions` call itself fails — a Subsonic-level error, or a non-2xx HTTP
 *   status such as a 404 from a server old enough not to implement the command — still negotiates
 *   to `isOpenSubsonic == true`, just with no extensions. Collapsing this into the same result as
 *   the first case would throw away information `ping` already established independently: this
 *   server *is* OpenSubsonic, it just would not (or could not) answer this one question.
 *
 * A genuine transport failure — no connection, a timeout, an unparseable body — is neither of
 * those. It is not a member of [SubsonicException] at all (see that type's own documentation for
 * why that boundary is structural, not a convention this class has to remember to respect), so it
 * is not caught by either `catch` clause below and simply propagates out of [negotiate] unchanged,
 * from *either* the `ping` call or the `getOpenSubsonicExtensions` call. "The server said no" and
 * "we could not ask" are different failures, and only the first justifies degrading to a
 * [ServerCapabilities] value instead of throwing.
 */
class CapabilityNegotiator(private val client: SubsonicClient) {

  /**
   * Runs the three-tier negotiation described in the class documentation and returns the result.
   *
   * Never catches anything around the `ping` call itself: a [SubsonicException] there (e.g. wrong
   * credentials) or a transport failure both propagate unchanged, exactly as calling
   * [SubsonicClient.ping] directly would. Negotiation only begins interpreting failures as "no
   * extensions" once `ping` has already succeeded and reported OpenSubsonic support.
   */
  suspend fun negotiate(): ServerCapabilities {
    val info = client.ping()
    if (!info.isOpenSubsonic) {
      return ServerCapabilities(isOpenSubsonic = false, extensions = emptyMap())
    }

    // Deliberately two catch clauses, not one `catch (e: SubsonicException)`: a sealed
    // *interface* is not itself a Throwable subtype, so the JVM will not allow catching it
    // directly (see SubsonicException's own documentation). SubsonicErrorException and
    // SubsonicHttpException are the sealed hierarchy's only two permitted members, so this pair is
    // the exhaustive equivalent of `catch (e: SubsonicException)` for a JVM that cannot express
    // that directly. Anything else — java.io.IOException, kotlinx.serialization.SerializationException
    // — is structurally incapable of matching either clause and propagates unchanged.
    val extensions =
      try {
        client.getOpenSubsonicExtensions()
      } catch (e: SubsonicErrorException) {
        emptyMap()
      } catch (e: SubsonicHttpException) {
        emptyMap()
      }

    return ServerCapabilities(isOpenSubsonic = true, extensions = extensions)
  }
}
